### PR TITLE
T 1532290: Serialize concurrent token refreshes to prevent one-time refresh token double-spend

### PR DESCRIPTION
**Pull request by vishal-mb**
- [x] I've read, understood, and done my best to follow the [*CONTRIBUTING guidelines*](https://github.com/mindbody/conduit/blob/master/CONTRIBUTING.md).

This pull request includes (pick all that apply):

- [x] Bugfixes
- [ ] New features
- [ ] Breaking changes
- [ ] Documentation updates
- [x] Unit tests
- [ ] Other

### Summary
`OAuth2RequestPipelineMiddleware`'s refresh coordination is check-then-act: `isRefreshTokenLockedFor` → `lockRefreshToken` with no atomicity, so two middlewares sharing a token store can both observe "unlocked" and both spend the same one-time refresh token. Against IdentityServer (`OneTimeOnly`, 10s consumed-grace) the straggler gets `invalid_grant`, the middleware deletes the token, and the user is force-logged out. This is a major contributor to Business App's daily forced logouts (same refresh token consumed 2–4× within seconds, server-side).

### Implementation
- New internal `OAuth2RefreshClaimRegistry`: atomic per-key claim under an `NSLock`, keyed by `tokenIdentifierFor` + client/authorization. `tryClaim` returns an ownership generation; stale claims (past the relinquish interval) are stealable so a hung refresh can't wedge refreshes; `release`/`owns` are generation-scoped so a superseded holder can't clear or clobber its successor's state (token store, persisted lock).
- Coordinated path (`refreshTokenAndResume`/`performRefresh`): winner re-reads the stored token after claiming and refreshes that (never a pre-claim snapshot), persists before releasing; losers reuse the persisted fresh token or wait briefly and re-enter. Waits are capped at 0.5s and once-guarded (`waitForRefresh` can fire from both timeout and notification when they race).
- Gated behind `OAuth2RequestPipelineMiddleware.refreshClaimCoordinationEnabled` (static, default `true`) as a kill switch; when disabled, the original refresh code runs unchanged — zero diff lines vs `main` in that region.

### Test Plan
- `OAuth2MiddlewareConcurrentRefreshTests` — deterministic reproduction of the race through unmodified production code (red before the fix: 2 grants spending the same token); asserts coalescing to one grant with both callers served, verifies a definitive `clientFailure` deletes the token before waking a waiter and prevents a second grant, and confirms disabling the flag restores the legacy double-spend behavior.
- `OAuth2MiddlewareRefreshClaimPathTests` — superseded holders (failure and success completions) can't delete/overwrite a successor's token or unlock its lock; winner-coalesce and loser fast-path branches serve a freshly persisted token with zero grants.
- `OAuth2RefreshClaimRegistryTests` — claim atomicity (1 winner of 32 concurrent), stale-steal, generation-scoped release/ownership.
- All 12 tests green; SwiftLint clean.

## Security Impact
- [x] No security impact

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: Sources/Conduit/Auth/OAuth2RequestPipelineMiddleware.swift [hunk 1]
```diff
@@ -103,6 +116,105 @@ public struct OAuth2RequestPipelineMiddleware: RequestPipelineMiddleware {
...
+                    tokenStorage.unlockRefreshTokenFor(client: clientConfiguration, authorization: authorization)
+                }
+                OAuth2RefreshClaimRegistry.shared.release(key: claimKey, generation: claimGeneration)
+                completion(.error(error))
+            case .value(let newToken):
+                logger.info("Successfully refreshed token")
+                logger.debug("New token issued: \(newToken)")
+                // Persist before releasing the claim or waking waiters; a superseded holder must not
+                // overwrite the successor's newer token or unlock its lock.
+                if OAuth2RefreshClaimRegistry.shared.owns(key: claimKey, generation: claimGeneration) {
```
Comment (AnshulMindbody) [inline Sources/Conduit/Auth/OAuth2RequestPipelineMiddleware.swift:205]: One edge case here looks like it can still log the user out. Imagine refresh A reaches the server and succeeds, but its callback is delayed long enough for the claim to expire. Refresh B can then steal the claim, retry the old refresh token, receive `invalid_grant`, and delete the stored token. When the valid response from A finally arrives, A no longer owns the claim, so the newly rotated token is never saved. Could we keep the earlier result until a newer refresh has successfully committed, and avoid deleting the token while the earlier request is still in flight?

File: Sources/Conduit/Auth/TokenStorage/Coordination/OAuth2RefreshClaimRegistry.swift [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+        nextGeneration += 1
+        claims[key] = Claim(generation: nextGeneration, expiresAt: Date().addingTimeInterval(staleAfter))
+        return nextGeneration
+    }
+
+    /// Returns whether `generation` still owns the claim; gates completion side effects of a possibly-superseded refresh.
+    func owns(key: String, generation: UInt64) -> Bool {
+        lock.lock()
+        defer { lock.unlock() }
+        return claims[key]?.generation == generation
```
Comment (AnshulMindbody) [inline Sources/Conduit/Auth/TokenStorage/Coordination/OAuth2RefreshClaimRegistry.swift:41]: I think there is still a race between checking ownership and updating the token store. `owns` releases the registry lock before the middleware stores or removes a token and releases the persisted refresh lock. In that gap the claim can expire and be stolen, after which the older refresh can overwrite the newer token or unlock the newer refresh lock. Could we atomically move the claim into a non-stealable committing state before these storage operations, then release it after they finish?

File: Sources/Conduit/Auth/OAuth2RequestPipelineMiddleware.swift [hunk 3]
```diff
@@ -103,6 +116,105 @@ public struct OAuth2RequestPipelineMiddleware: RequestPipelineMiddleware {
...
+            case .error(let error):
+                logger.warn("There was an error refreshing the token")
+                // A superseded holder must not delete the successor's token or unlock its lock.
+                if OAuth2RefreshClaimRegistry.shared.owns(key: claimKey, generation: claimGeneration) {
+                    if case OAuth2Error.clientFailure = error {
+                        self.tokenStorage.removeTokenFor(client: self.clientConfiguration, authorization: self.authorization)
+                    }
+                    tokenStorage.unlockRefreshTokenFor(client: clientConfiguration, authorization: authorization)
+                }
+                OAuth2RefreshClaimRegistry.shared.release(key: claimKey, generation: claimGeneration)
```
Comment (AnshulMindbody) [inline Sources/Conduit/Auth/OAuth2RequestPipelineMiddleware.swift:198]: I do not think waiting requests should retry after the first refresh reports a network, server, or internal failure. The token server may already have consumed the one-time refresh token even if its response was lost. Releasing the claim here wakes a waiter, which sends the same token again and may receive `invalid_grant` and delete the session. Could the in-process waiters receive the same refresh result instead? A concurrent failure test would help cover this case.

**Review by AnshulMindbody — COMMENTED** [review]
I found a few concurrency edge cases that may still allow the same one-time refresh token to be spent more than once.

File: Sources/Conduit/Auth/OAuth2RequestPipelineMiddleware.swift (hunk 1 above)
Comment (vishal-mb) [reply Sources/Conduit/Auth/OAuth2RequestPipelineMiddleware.swift:205]: Good catch. This ordering is real, including the nuance that main could re-persist the late rotated token after the successor deleted it. We are retaining it as the accepted stale-steal limitation: both the in-process claim and the pre-existing persisted lock are intentionally stealable after the 30-second relinquish interval so a lost completion cannot wedge refreshes indefinitely. Once superseded, the late holder cannot distinguish a successor failure from a successor that committed a newer token. Preserving the older result safely needs delayed multi-generation arbitration plus a conditional or generation-aware token-store write; cross-process correctness likewise needs store-level CAS support, outside this middleware-only change. The production race addressed here occurs within milliseconds, before claim expiry, so no code change is planned for this bounded case.

File: Sources/Conduit/Auth/TokenStorage/Coordination/OAuth2RefreshClaimRegistry.swift (hunk 2 above)
Comment (vishal-mb) [reply Sources/Conduit/Auth/TokenStorage/Coordination/OAuth2RefreshClaimRegistry.swift:41]: Agreed that owns is a snapshot and the check-to-side-effect gap is technically real. We are retaining it as an accepted residual: it requires the claim to be stale and a steal to land in the short interval after the ownership snapshot, versus the entire tail of a slow grant before the gate. A committing state is better than holding the registry mutex through storage I/O, but it only closes the in-process case; if non-expiring it can wedge the same key when a consumer store blocks, and if leased it retains another expiry boundary. It also cannot make the persisted cross-process lock generation-aware. Full atomicity therefore needs an OAuth2TokenStore CAS or generation protocol change, which is explicitly outside this PR.

File: Sources/Conduit/Auth/OAuth2RequestPipelineMiddleware.swift (hunk 3 above)
Comment (vishal-mb) [reply Sources/Conduit/Auth/OAuth2RequestPipelineMiddleware.swift:198]: Thanks; I added the requested concurrent definitive-failure coverage in a10843b. The test parks the winning grant, registers a waiter, returns clientFailure, and verifies that the token is deleted before the waiter resumes: both callers fail and the factory records only the original RT_OLD grant. We are intentionally retaining retries for ambiguous network, server, no-response, and internal failures. IdentityServer permits reuse inside its 10-second consumed-token grace, so the immediate claim-serialized retry can recover a rotated token whose first response was lost. Propagating the failure would remove that rescue path, while a response arriving after the grace period is unrecoverable from the stored credential either way. The claim permits only one successor grant at a time; a definitive clientFailure removes the token before release, so waiters fail fast without another spend.

**Review by AnshulMindbody — APPROVED** [review]
LGTM
